src/sdam/topology_description.ts

```diff
@@ -139,23 +140,51 @@ export class TopologyDescription {
    * Returns a new TopologyDescription based on the SrvPollingEvent
    * @internal
    */
-  updateFromSrvPollingEvent(ev: SrvPollingEvent): TopologyDescription {
-    const newAddresses = ev.addresses();
-    const serverDescriptions = new Map(this.servers);
-    for (const address of this.servers.keys()) {
-      if (newAddresses.has(address)) {
-        newAddresses.delete(address);
-      } else {
-        serverDescriptions.delete(address);
+  updateFromSrvPollingEvent(ev: SrvPollingEvent, srvMaxHosts = 0): TopologyDescription {
+    /** The SRV addresses defines the set of addresses we should be using */
+    const incomingHostnames = ev.hostnames();
+    const currentHostnames = new Set(this.servers.keys());
+
+    const hostnamesToRemove = new Set<string>();
+    for (const hostname of currentHostnames) {
+      if (!incomingHostnames.has(hostname)) {
+        // If the SRV Records no longer include this hostname
+        // we have to stop using it
+        hostnamesToRemove.add(hostname);
+      }
```

Comment:
I think if we seed `hostnamesToAdd` with the incomingHostnames and add an else loop to remove the matching names here, you'll be left with the correct set of `hostnamesToAdd` without needing the second loop below